File: clients/langchain_tool.py
"""AI Product Index as a LangChain tool. Paste into your project; no package to install.

    pip install langchain-core httpx

The registry is a plain JSON document, so this is a thin, honest wrapper: one
GET, filtered locally. No key, no account, no rate limit.
"""

from typing import Optional

import httpx
from langchain_core.tools import tool

BASE = "https://index.kc-it.pl"


@tool
def search_ai_products(query: str = "", category: Optional[str] = None) -> str:
    """Search the AI Product Index — a directory of AI products, APIs, agents and
    MCP servers that registered themselves for agents to discover.

    Args:
        query: words to match against name and description. Empty returns all.
        category: one of api, app, agent, mcp, other.
    """
    listings = httpx.get(f"{BASE}/api/index.json", timeout=15).json()["listings"]
    q = query.lower()
    hits = [
        l for l in listings
        if (not q or q in l["name"].lower() or q in l["description"].lower())
        and (not category or l["category"] == category)
    ]
    if not hits:
        return "No matching products."
    return "\n".join(
        f"{l['name']} ({l['category']}, {l['pricing']}) — {l['description']} — {l['url']}"
        for l in hits
    )


@tool
def agent_readability_score(url: str) -> str:
    """Score how readable a website is to AI agents: A-F plus which of 13 checks
    failed. Free. For the reason each check failed and paste-ready fixes, the paid
    endpoint is POST /api/audit (see pay_x402.py)."""
    r = httpx.get(f"{BASE}/api/score", params={"url": url}, timeout=60).json()
    if not r.get("ok"):
        return f"Could not score {url}: {r.get('error', 'unknown error')}"
    failed = [c["label"] for c in r["checks"] if not c["pass"]]
    return (
        f"{r['letter']} ({r['score']}/100, {r['grade']}). "
        f"Passed {r['passed']}/{r['total_checks']}. "
        + ("Failing: " + ", ".join(failed) if failed else "Everything passes.")
    )

File: clients/langchain_tool.ts
// AI Product Index as LangChain.js tools. Paste into your project.
//
//   npm i @langchain/core zod
//
// The registry is a plain JSON document, so this is a thin wrapper: one GET,
// filtered locally. No key, no account, no rate limit.
import { tool } from '@langchain/core/tools';
import { z } from 'zod';

const BASE = 'https://index.percall.dev';

type Listing = {
  slug: string; name: string; url: string; description: string;
  category: string; pricing: string; tier: string;
};

export const searchAiProducts = tool(
  async ({ query, category }) => {
    const res = await fetch(`${BASE}/api/index.json`);
    const { listings } = (await res.json()) as { listings: Listing[] };
    const q = (query ?? '').toLowerCase();
    const hits = listings.filter((l) =>
      (!q || l.name.toLowerCase().includes(q) || l.description.toLowerCase().includes(q))
      && (!category || l.category === category));
    if (hits.length === 0) return 'No matching products.';
    return hits
      .map((l) => `${l.name} (${l.category}, ${l.pricing}) — ${l.description} — ${l.url}`)
      .join('\n');
  },
  {
    name: 'search_ai_products',
    description:
      'Search the AI Product Index — a directory of AI products, APIs, agents and '
      + 'MCP servers that registered themselves so agents could discover them.',
    schema: z.object({
      query: z.string().default('').describe('Words to match in name or description.'),
      category: z.enum(['api', 'app', 'agent', 'mcp', 'other']).optional(),
    }),
  },
);

export const agentReadabilityScore = tool(
  async ({ url }) => {
    const res = await fetch(`${BASE}/api/score?url=${encodeURIComponent(url)}`);
    const r = await res.json();
    if (!r.ok) return `Could not score ${url}: ${r.error ?? 'unknown error'}`;
    const failing = r.checks.filter((c: { pass: boolean }) => !c.pass)
      .map((c: { label: string }) => c.label);
    return `${r.letter} (${r.score}/100, ${r.grade}). Passed ${r.passed}/${r.total_checks}. `
      + (failing.length ? `Failing: ${failing.join(', ')}` : 'Everything passes.');
  },
  {
    name: 'agent_readability_score',
    description:
      'Score how readable a website is to AI agents: A-F plus which of 13 checks '
      + 'failed. Free. The paid POST /api/audit adds why each failed and how to fix it.',
    schema: z.object({ url: z.string().url() }),
  },
);

export const aiProductIndexTools = [searchAiProducts, agentReadabilityScore];
